### DISCUSSION TITLE
Rust Bitcoin Summit 2024

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#3112] [General]
Discussion thread for things that happened at the Rust Bitcoin Summit in Nashville July 2024.

**Comment by tcharding** [comment]
Day 1, we talked productively and reached a few "gentleman's agreements":

- Ecosystem to stay on `rust-bitcoin v0.32.0` until we get `primitives v1.0` released. 
- Consider doing some sort of compat layer between `v0.33.0`/`primtives` and `v0.32.0`.
- Consider releasing `addresses v0.0.1` with nothing but a dumb type with `to/from_bytes`.
- Language bindings for `rust-bitcoin` and `bdk` to go in the `github.com/bitcoindevkit` org.

**Comment by DanGould** [comment]
We've also stated we're looking for help with a few things. I imagine this list is going to grow. 

- [ ] Someone to support SIMD / crypto optimizations in rust-bitcoin for bitcoin-hashes and bech32
- [ ] Write an async version of [`minreq`](https://crates.io/crates/minreq) ([see issue](https://github.com/neonmoe/minreq/issues/97))
- [ ] Help with (https://github.com/tcharding/rust-bitcoind-json-rpc) (or is this already done?)
- [x] Help writing [HPKE](https://crates.io/crates/hpke) and [OHTTP](https://crates.io/ohttp) crates that depend on rust-secp256k1 to implement DHKEM(secp256k1, HKDF-SHA256) with ChaCha20-Poly1305 AEAD. These are used as a default Tor alternative in `payjoin`. More details can be found [here](https://github.com/payjoin/rust-payjoin/issues/216). edit: done! [bitcoin_hpke](https://crates.io/crates/bitcoin-hpke) and [bitcoin-ohttp](https://crates.io/crates/bitcoin-ohttp). The hpke crate could still rely on bitcoin-hashes for sha256

If anyone wants to help or has leads for those who might be able to help please do share

**Reply by Kixunil** [reply]
> * Someone to support SIMD / crypto optimizations in rust-bitcoin for bitcoin-hashes and bech32

Are there any known identified hot paths? We got SHA256 covered already on x86 (arm would be great too though).

> * Write an async version of [`minreq`](https://crates.io/crates/minreq)

I had the crazy idea of writing `push_decode` impls for HTTP but maybe it's not that crazy after all?

**Reply by apoelstra** [reply]
For sha256 I think we have a stronger goal -- to use the SSE4/SHA CPU acceleration when available. I don't think we've done that yet.

Then the other hashes presumably could be SIMD-optimized in much the same way as sha2. Is that worth the effort? I don't know. I think if somebody else was willing to put the effort in, we'd appreciate it :P. In Bitcoin sha2 is used everywhere, ripemd160 quite rarely, and other hashes almost never.

Then as for bech32 and poly1305 -- we didn't know whether SIMD offered any value there, but if it does, that could be a big performance boost.

**Reply by clarkmoody** [reply]
The bech32 encode / decode operations are already very simple and fast. I'm not sure the additional complexity would be worth it. The error-correction logic is much slower, but that's not the primary use-case.

**Reply by DanGould** [reply]
It looks like some progress has been made by @BEULAHEVANJALIN toward [async-minreq](https://github.com/BEULAHEVANJALIN/async-minreq/commits/master/)

**Comment by DanGould** [comment]
As of yesterday we're also looking for help improving [bip21](https://github.com/Kixunil/bip21) and specifically addressing [query parameters' `Extras` implementation composibility](https://github.com/Kixunil/bip21/issues/22). We discussed moving the module into the `rust-bitcoin` org and supporting BlueMatt's [proposed bip21 update](https://github.com/bitcoin/bips/pull/1555), and finding help for the implementations.

**Comment by DanGould** [comment]
Another thing to seek help with is a Rust version of HWI.

**Reply by oleonardolima** [reply]
IIRC someone mentioned this one: https://github.com/wizardsardine/async-hwi. Maybe it could be a good candidate itself or a good start for a fork.

**Reply by Kixunil** [reply]
They have a trait. Maybe impl the trait for Rust-native types?

**Comment by tcharding** [comment]
it was mentioned in parsing that our changelogs are hard to use because:

1. We sometimes lose the point release changelogs on master
2. They are spread across crates

We should fix (1) but not sure what we can do about (2).

**Reply by Kixunil** [reply]
I think at best we should link the changelogs of dependencies. If I needed the changelog then I'd be really annoyed to read stuff from `bitcoin` if I'm only looking for `units`.

**Comment by tcharding** [comment]
FYI here are notes taken by @notmandatory during the summit:

https://hackmd.io/@notmandatory/r1tcjErKR

**Reply by Kixunil** [reply]
What is "invert secp traits for impl by default prod or other secp c or non-c"?

**Reply by apoelstra** [reply]
I don't think we came to an agreement on what that might actually mean. Some people wanted to be able to swap the C code out for other code. I argued that "letting users replace all the hard crypto" was in conflict with "making a library that was hard to misuse and not designed for roll-your-own-crypto stuff".

But OTOH I have secp256k1-zkp, Lloyd Fournier has secp256k1-fun, etc., and currently if you want to use these libraries you basically need to link in the whole library alongside all of libsecp256k1, even though they have 99% overlap, which sucks.

Relatedly, such a solution might let people use Rust code if they don't want to link with C ... but I think we have decided to just solve that directly by doing a non-C implementation of the safe/easy stuff.

But I don't think we have a solution.

**Reply by Kixunil** [reply]
Oh, I see. Yes, I think we solve most of that with easy Rust impl but also I think we can provide `PublicKey::from_bytes_assume_valid` which people could use instead of secp256k1 calls and implement their own checking or other operations using another library. I think the zkp variant is quite serious reason to do this.

**Comment by BEULAHEVANJALIN** [comment]
@DanGould I've implemented async version of minreq https://github.com/BEULAHEVANJALIN/async-minreq

**Reply by DanGould** [reply]
Fantastic! This has been a want for a super long time. Thanks for picking it up.

I wonder how much it would take to prove the concept in one of the rust-bitcoin projects. [payjoin-cli](https://github.com/payjoin/rust-payjoin/tree/master/payjoin-cli) uses reqwest, because we're in an async runtime and minreq didn't work there, and I know @TheBlueMatt had some interest in an async-minreq as well but not sure in what context. The corerpc crate uses minreq, but I'm not sure if @tcharding had to work around the blocking nature of minreq in deciding to use it or not.

**Reply by tcharding** [reply]
`corepc-client` is only a blocking client. I don't personally have any intention on adding async to it.

**Reply by tcharding** [reply]
cc @jamillambert because he is playing around with writing a HTTP crate as well at the moment.

**Reply by TheBlueMatt** [reply]
It would be nice if we had a single async+sync minreq-type thing, with some standard proc-macro that builds async+sync versions of simple methods. That would make things like `corepc-client` fairly easy to adapt to dual-sync-async and make our ecosystem much less fragmented. If we have a single async+sync minreq-type client (with the ability to stub out the entire request in wasm) then we could all use a single http dependency rather than having a million dependencies used across the ecosystem.

**Reply by BEULAHEVANJALIN** [reply]
> It would be nice if we had a single async+sync minreq-type thing, with some standard proc-macro that builds async+sync versions of simple methods. That would make things like `corepc-client` fairly easy to adapt to dual-sync-async and make our ecosystem much less fragmented. If we have a single async+sync minreq-type client (with the ability to stub out the entire request in wasm) then we could all use a single http dependency rather than having a million dependencies used across the ecosystem.


This is an interesting idea. Let me work on this further
